lib/zstd_compress.c

```diff
@@ -1504,118 +1689,195 @@ static size_t ZSTD_BtFindBestMatch_selectMLS_extDict (
 }
 
 
-/* ***********************
-*  Hash Chain
-*************************/
-
-#define NEXT_IN_CHAIN(d, mask)   chainTable[(d) & mask]
 
-/* Update chains up to ip (excluded)
-   Assumption : always within prefix (ie. not within extDict) */
+/* *******************************
+*  Greedy parser
+*********************************/
 FORCE_INLINE
-U32 ZSTD_insertAndFindFirstIndex (ZSTD_CCtx* zc, const BYTE* ip, U32 mls)
+void ZSTD_compressBlock_greedy_generic(ZSTD_CCtx* ctx,
+                                     const void* src, size_t srcSize)
 {
-    U32* const hashTable  = zc->hashTable;
-    const U32 hashLog = zc->params.cParams.hashLog;
-    U32* const chainTable = zc->chainTable;
-    const U32 chainMask = (1 << zc->params.cParams.chainLog) - 1;
-    const BYTE* const base = zc->base;
-    const U32 target = (U32)(ip - base);
-    U32 idx = zc->nextToUpdate;
-
-    while(idx < target) {
-        size_t const h = ZSTD_hashPtr(base+idx, hashLog, mls);
-        NEXT_IN_CHAIN(idx, chainMask) = hashTable[h];
-        hashTable[h] = idx;
-        idx++;
-    }
+    seqStore_t* seqStorePtr = &(ctx->seqStore);
+    const BYTE* const istart = (const BYTE*)src;
+    const BYTE* ip = istart;
+    const BYTE* anchor = istart;
+    const BYTE* const iend = istart + srcSize;
+    const BYTE* const ilimit = iend - 8;
+    const BYTE* const base = ctx->base + ctx->dictLimit;
 
-    zc->nextToUpdate = target;
-    return hashTable[ZSTD_hashPtr(ip, hashLog, mls)];
-}
+    const U32 maxSearches = 1 << ctx->params.cParams.searchLog;
+    const U32 mls = ctx->params.cParams.searchLength;
 
+    /* init */
+    U32 rep[ZSTD_REP_INIT];
+    for (U32 i=0; i<ZSTD_REP_INIT; i++)
```

Comment:
I found this construction in a number of places :
`for (U32 i=0; ...`

While I really do like it, and would like to use it,
it's unfortunately incompatible with non-C99 compilers.

Yes, the year is 2016, but it's nonetheless too soon to drop support for these compilers,
the goal being wide portability.

Therefore prefer : 
`{ U32 i; for (i=0; ... `
which is C89 compatible.

It requires one more bracket. Still, portability is more important.

_Note_ : I made the associated modifications in ea63bb7b5e41db6143905e7c8a0ff2385a2ff677 .

_Note 2_ : maybe we need a test to ensure this compatibility is respected. Currently, all compilations in `Makefile` are done with `-std=c99` compilation flag.